### PR TITLE
Add template meta programming lecture materials

### PR DESCRIPTION
**Pull request by ivankochin**

### CODE REVIEW

File: 12_template_meta_programming/12_00_compile_time_computations/12_00_00_constant_expressions.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// What are constant expressions and where they may be used?
+// How to check if an expression is constant?
+
+#include <stddef>
```
Comment (ivankochin) [inline 12_template_meta_programming/12_00_compile_time_computations/12_00_00_constant_expressions.cpp]:
```suggestion
#include <cstddef>

```

File: 12_template_meta_programming/12_00_compile_time_computations/12_00_01_factorial.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Factorial computation during the compilation process
+// Also may be computated inside constexpr function after C++11
+
+#include <stddef>
```
Comment (ivankochin) [inline 12_template_meta_programming/12_00_compile_time_computations/12_00_01_factorial.cpp]:
```suggestion
#include <сstddef>
```

File: 12_template_meta_programming/12_02_integral_constants/12_02_00_bool_constants.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    // since C++17
+    template <bool B>
+    using bool_constant = integral_constant<bool, B>;
+
+    using true_type = bool_constant<true>;
+    using false_type = bool_constant<false>;
+
+} // namespace my
+
+int main() {
```
Comment (ivankochin) [inline 12_template_meta_programming/12_02_integral_constants/12_02_00_bool_constants.cpp]: Rework the example

File: 12_template_meta_programming/12_03_standard_traits/12_03_00_is_same.cpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    struct is_same : public std::false_type { };
+
+    template <typename T>
+    struct is_same<T, T> : public std::true_type { };
+
+    // sinse C++14
+    template<typename T, typename U>
+    using is_same_t = is_same<T, U>::type;
+
+    // sinse C++14
```
Comment (ivankochin) [inline 12_template_meta_programming/12_03_standard_traits/12_03_00_is_same.cpp]:
```suggestion
    // sinse C++17
```

File: 12_template_meta_programming/12_03_standard_traits/12_03_00_is_same.cpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    template <typename T, typename U>
+    struct is_same : public std::false_type { };
+
+    template <typename T>
+    struct is_same<T, T> : public std::true_type { };
+
+    // sinse C++14
+    template<typename T, typename U>
+    using is_same_t = is_same<T, U>::type;
```
Comment (ivankochin) [inline 12_template_meta_programming/12_03_standard_traits/12_03_00_is_same.cpp]:
```suggestion

```

File: 12_template_meta_programming/12_03_standard_traits/12_03_04_is_nothrow_move_constructible.cpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Additional example
+
+#include <iostream>
+#include <utility>
+
+namespace my {
+
+    template<typename T>
+    struct is_nothrow_move_constructible :
+        std::integral_constant<bool, noexcept(T(std::declval<T&&>()))>
```
Comment (ivankochin) [inline 12_template_meta_programming/12_03_standard_traits/12_03_04_is_nothrow_move_constructible.cpp]:
```suggestion
        std::integral_constant<bool, noexcept(T(std::declval<T>()))>

```

File: 12_template_meta_programming/12_04_conditional/12_02_00_conditional_example.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+template<bool, typename, typename U>
+struct conditional {
+    using type = U;
+};
+
+template<typename T, typename U>
+struct conditional<true, T, U> {
+    using type = T;
+};
+
```
Comment (ivankochin) [inline 12_template_meta_programming/12_04_conditional/12_02_00_conditional_example.cpp]: Implement `conditional_t`

File: 12_template_meta_programming/12_02_tag_dispatching/12_02_00_sorted_find.cpp [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    return sorted_find_impl(begin, end, key, category{});
+}
+
+int main() {
+    std::vector<int> v{};
+    std::set<int> s{};
+
+    sorted_find(v.begin(), v.end(), 1);
+    sorted_find(s.begin(), s.end(), 1);
+}
```
Comment (ivankochin) [inline 12_template_meta_programming/12_02_tag_dispatching/12_02_00_sorted_find.cpp]:
```suggestion
}

```

File: 12_template_meta_programming/12_00_compile_time_computations/12_00_00_constant_expressions.cpp [hunk 9]
```diff
@@ -0,0 +1,18 @@
+// What are constant expressions and where they may be used?
+// How to check if an expression is constant?
```
Comment (rarutyun) [inline 12_template_meta_programming/12_00_compile_time_computations/12_00_00_constant_expressions.cpp]:
```suggestion
```

**Pull request merged**